### PR TITLE
User Registration / Navigation

### PR DESCRIPTION
**Pull request by DIzosimov**
# User Registration and Navigation
[Link to PT](https://www.pivotaltracker.com/story/show/168829428)
[Link to PT](https://www.pivotaltracker.com/story/show/168829436)

## PR content
In this PR we are adding Signup and NavBar functionality with the help of the gem Devise.

## This PR contains
- Configurations for devise
- Login Views
- Signup Views
- Signup Functionality
- Feature Tests
- A user model
- User model specs
- User FactoryBot
- Adds NavBar Views
- Adds NavBar Functionality

## What I have learned
- How to configure devise
- How to resolve merge conflicts
- How to adapt to peer pull requests and changes
- How to write feature tests
- How to create specs
- How to add signup functionality to our rails application
- How controllers work (data flow)

## Images

Screenshot 2019-10-01 at 15 56 03
Screenshot 2019-10-01 at 15 56 57

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,3 +1,10 @@
 Rails.application.routes.draw do
-  root controller: :product, action: :index
+  get 'products/show'
+  devise_for :users
+
+  get 'landing/index'
```
Comment (piavW) [inline config/routes.rb]: Proper syntax - remove get '.../..' Look at coursematerial about Routing /Faraz's demo.

File: features/sign_up.feature [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Feature: Sign Up
+As a user
```
Comment (piavW) [inline features/sign_up.feature]: indentation for line 2-4

**Review by piavW — APPROVED** [review]
See comments

**Review by yasmineezequiel — APPROVED** [review]

**Comment by DavveDavve** [comment]
Screenshot 2019-10-01 at 15 56 57
Screenshot 2019-10-01 at 15 56 03

File: features/step_definitions/assertion_steps.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Then("I should see {string}") do |content|
+  expect(page).to have_content content
+end
+
+Then("I should be on Login page") do
+  visit new_user_session_path
+end
+
+Then("I should be on Sign up page") do
+  visit new_user_registration_path
```
Comment (oliverochman) [inline features/step_definitions/assertion_steps.rb]: is this an assertion?

File: app/controllers/landing_controller.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+class LandingController < ApplicationController
+  def index
+  end
+end
```
Comment (faraznaeem) [inline app/controllers/landing_controller.rb]: Is this controller still neccessary?

File: features/step_definitions/assertion_steps.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Then("I should see {string}") do |content|
+  expect(page).to have_content content
+end
+
+Then("I should be on Login page") do
+  visit new_user_session_path
```
Comment (oliverochman) [inline features/step_definitions/assertion_steps.rb]: is this an assertion?

File: features/step_definitions/basic_steps.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+Given("I visit the landing page") do
+  visit root_path
+end
+
+Given("I click {string}") do |element|
+  click_on element
+end
+
+Given("I visit the menu page") do
+  visit products_path
```
Comment (oliverochman) [inline features/step_definitions/basic_steps.rb]: whats the difference between the root_path and this path?

File: features/step_definitions/basic_steps.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+end
+
+When("I fill in {string} with {string}") do |field, content|
+  fill_in field, with: content
+end
+
+Given("the following products exists") do |table|
```
Comment (oliverochman) [inline features/step_definitions/basic_steps.rb:13]: why is this step added in this feature?

File: config/routes.rb [hunk 8]
```diff
@@ -1,3 +1,10 @@
...
+  
+  resources :products
```
Comment (oliverochman) [inline config/routes.rb]: please clean this file up from redundant routes

File: app/views/devise/unlocks/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend unlock instructions</h2>
+
+<%= form_for(resource, as: resource_name, url: unlock_path(resource_name), html: { method: :post }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
```
Comment (oliverochman) [inline app/views/devise/unlocks/new.html.erb]: please remove all genereated devise views that u do not change

File: features/sign_up.feature (hunk 2 above)
Comment (faraznaeem) [inline features/sign_up.feature]: Please mind the indentation

File: app/views/devise/registrations/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,34 @@
+<h2>Sign up</h2>
+
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+
+  <div class="field">
```
Comment (oliverochman) [inline app/views/devise/registrations/new.html.erb]: please write this file in haml

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+      it { is_expected.to have_db_column :name }
+      it { is_expected.to have_db_column :email }
+      it { is_expected.to have_db_column :encrypted_password }
+    end
+    describe 'Validations' do
+      it { is_expected.to validate_presence_of :name }
+      it { is_expected.to validate_uniqueness_of :name }
+      it { is_expected.to validate_presence_of :email }
+    end
+    describe 'Factory' do
```
Comment (faraznaeem) [inline spec/models/user_spec.rb]: Add lines

File: app/controllers/landing_controller.rb [hunk 12]
```diff
@@ -0,0 +1,4 @@
+class LandingController < ApplicationController
```
Comment (oliverochman) [inline app/controllers/landing_controller.rb]: why did you add a LandingController when u already have the products controller?

File: app/controllers/application_controller.rb [hunk 13]
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
+
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
+    devise_parameter_sanitizer.permit(:account_update, keys: [:name])
```
Comment (oliverochman) [inline app/controllers/application_controller.rb]: why do you sanitize for account_update in this PR? 
out of scope

**Review by oliverochman — CHANGES_REQUESTED** [review]
I will let this through: In the future, you should only have one PT story per PR

File: features/step_definitions/assertion_steps.rb (hunk 5 above)
Comment (faraznaeem) [inline features/step_definitions/assertion_steps.rb]: please make use of expectations when writing the assertions:
`expect(current_path).to eq new_user_session_path`

**Review by faraznaeem — CHANGES_REQUESTED** [review]

File: features/step_definitions/basic_steps.rb (hunk 6 above)
Comment (DIzosimov) [reply features/step_definitions/basic_steps.rb]: Changed according to comments.

File: config/routes.rb (hunk 8 above)
Comment (clalil) [reply config/routes.rb]: Changed according to comments.

File: features/step_definitions/assertion_steps.rb (hunk 5 above)
Comment (clalil) [reply features/step_definitions/assertion_steps.rb]: Changed according to comments.

File: features/step_definitions/assertion_steps.rb (hunk 3 above)
Comment (clalil) [reply features/step_definitions/assertion_steps.rb]: Changed according to Faraz comments.

File: features/step_definitions/basic_steps.rb (hunk 7 above)
Comment (clalil) [reply features/step_definitions/basic_steps.rb:13]: Used for sign up and login feature.

File: spec/models/user_spec.rb (hunk 11 above)
Comment (clalil) [reply spec/models/user_spec.rb]: Changed according to comments.

File: app/views/devise/sessions/new.html.haml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+%h2 Log in
+= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f|
+  .field
```
Comment (oliverochman) [inline app/views/devise/sessions/new.html.haml]: please remove this file, no need to have it in the code base

File: app/views/devise/shared/_error_messages.html.haml [hunk 15]
```diff
@@ -0,0 +1,9 @@
+- if resource.errors.any?
```
Comment (oliverochman) [inline app/views/devise/shared/_error_messages.html.haml]: please remove this file, no need to have it in the code base

File: app/views/devise/shared/_links.html.haml [hunk 16]
```diff
@@ -0,0 +1,10 @@
+- if controller_name != 'sessions'
+  = link_to "Log in", new_session_path(resource_name)
```
Comment (oliverochman) [inline app/views/devise/shared/_links.html.haml]: please remove this file, no need to have it in the code base

File: app/views/layouts/application.html.haml [hunk 17]
```diff
@@ -7,4 +7,5 @@
     = stylesheet_link_tag    'application', media: 'all', data: {'turbolinks-track' => 'reload'}
     = javascript_include_tag 'application', data: {'turbolinks-track' => 'reload'}
   %body
+    = render 'layouts/nav'
```
Comment (oliverochman) [inline app/views/layouts/application.html.haml:10]: render flashes in this file

File: features/sign_up.feature [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+  As a user
+  In order to be able to finalise an order
+  I would like to be able to register an account
+
+  Background:
+    Given I visit the landing page
+    And I click "Login"
+    Then I should be on Login page
+    And I click "Sign up"
+    Then I should be on Sign up page
```
Comment (oliverochman) [inline features/sign_up.feature]: NEver have any assertions in the background step

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: features/sign_up.feature (hunk 2 above)
Comment (clalil) [reply features/sign_up.feature]: Changed accordingly.

File: app/views/devise/sessions/new.html.haml (hunk 14 above)
Comment (clalil) [reply app/views/devise/sessions/new.html.haml]: Don't we need it for our login feature?

File: app/views/devise/shared/_error_messages.html.haml (hunk 15 above)
Comment (clalil) [reply app/views/devise/shared/_error_messages.html.haml]: Changed accordingly.

File: app/views/devise/shared/_links.html.haml (hunk 16 above)
Comment (clalil) [reply app/views/devise/shared/_links.html.haml]: Don't we need it for loggging in?

File: app/views/layouts/application.html.haml (hunk 17 above)
Comment (clalil) [reply app/views/layouts/application.html.haml:10]: Changes made accordingly.

File: features/sign_up.feature (hunk 18 above)
Comment (clalil) [reply features/sign_up.feature]: Changes made accordingly.

File: app/views/devise/sessions/new.html.haml (hunk 14 above)
Comment (oliverochman) [reply app/views/devise/sessions/new.html.haml]: nope, try removing it and run the tests

File: app/views/devise/shared/_links.html.haml (hunk 16 above)
Comment (oliverochman) [reply app/views/devise/shared/_links.html.haml]: nope same here, try to run the tests without this file

File: features/sign_up.feature (hunk 18 above)
Comment (oliverochman) [reply features/sign_up.feature]:
```
Feature: Sign Up
  As a user
  In order to be able to finalise an order
  I would like to be able to register an account

  Background:
    Given I visit the landing page
    And I click "Login"
    And I click "Sign up"

  Scenario: Creating an account
    When I fill in "Name" with "John"
    And I fill in "Email" with "john@doe.com"
    And I fill in "Password" with "password"
    And I fill in "Password confirmation" with "password"
    And I click "Sign up"
    Then I should see "Hello, John"
```

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/views/devise/sessions/new.html.haml (hunk 14 above)
Comment (clalil) [reply app/views/devise/sessions/new.html.haml]: NO we CAN'T!!! If we do, our sign up function is removed along with it.

File: app/views/devise/shared/_links.html.haml (hunk 16 above)
Comment (clalil) [reply app/views/devise/shared/_links.html.haml]: Changes made according to your feedback.

File: features/sign_up.feature (hunk 18 above)
Comment (clalil) [reply features/sign_up.feature]: Changes made according to your feedback.

**Review by clalil — APPROVED** [review]

**Review by oliverochman — APPROVED** [review]

**Review by jonas-masharqa — APPROVED** [review]
